File: lesson_5_2/app/src/main/java/com/kalashnyk/denys/samplemvvmapp/repository/AppRepository.java
package com.kalashnyk.denys.samplemvvmapp.repository;

import com.kalashnyk.denys.samplemvvmapp.repository.database.dao.UserDao;
import com.kalashnyk.denys.samplemvvmapp.repository.database.entity.UserEntity;
import com.kalashnyk.denys.samplemvvmapp.repository.server.ServerCommunicator;

import java.util.List;
import java.util.Objects;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class AppRepository {
    private ServerCommunicator mServerCommunicator;
    private UserDao mDatabase;

    public AppRepository(ServerCommunicator serverCommunicator, UserDao database) {
        mServerCommunicator = serverCommunicator;
        mDatabase = database;
    }

    public Single<List<UserEntity>> getAllUsers() {
        return mServerCommunicator.getAllUsers()
                .flatMap(list -> {
                    mDatabase.insertList(Objects.requireNonNull(list.body()).getRecords());
                    return Single.just(mDatabase.getAll());
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<UserEntity> getUser(int id) {
        return mServerCommunicator.getUser(id)
                .map(userEntity -> {
                    UserEntity user = mDatabase.getById(id);
                    return user != null ? user : user;
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());

    }
}

File: lesson_7/app/src/main/java/com/kalashnyk/denys/kotlinsample/repository/AppRepository.kt
package com.kalashnyk.denys.kotlinsample.repository

import com.kalashnyk.denys.kotlinsample.repository.database.AppDatabase
import com.kalashnyk.denys.kotlinsample.repository.database.entity.UserEntity
import com.kalashnyk.denys.kotlinsample.repository.server.ServerCommunicator
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.*

class AppRepository(private val serverCommunicator: ServerCommunicator, private val mainDatabase: AppDatabase) {

    fun getAll(): Single<List<UserEntity>>? {
        return serverCommunicator.getAllUsers()
            .flatMap { list ->
                mainDatabase.userDao().insertList(Objects.requireNonNull(list.body())?.records!!)
                Single.just(mainDatabase.userDao().getAll())
            }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
    }

    fun getUser(id: Int): Single<UserEntity> {
        return serverCommunicator.getUser(id)
            .map {
                val user = mainDatabase.userDao().getById(id)
                user
            }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())

    }
}